### PR TITLE
Python 3.11

### PR DESCRIPTION
**Pull request by chassing**
Switch from Python 3.9 to 3.11

Depends on: 
* https://github.com/app-sre/container-images/pull/85
* https://github.com/app-sre/container-images/pull/86

For the reviewer:
* I added `@runtime_checkable` to some protocols due to different behavior in Python 3.11
* I adapted creating mocks in the tests due to different behavior in Python 3.11
* I replaced deprecated `imghdr` with `filetype`
* Bump `jenkins-job-builder` to allow installation of PyYAML 6.0.1 (https://github.com/yaml/pyyaml/pull/702) to mitigate `Cython 3.0` build errors

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -219,7 +219,7 @@ OpenShift templates can be found [here](/openshift/qontract-reconcile.yaml). In
 
 ## Installation
 
-This project targets Python version 3.9.x for best compatibility. Verify the Python3 version that your shell is using with `python3 --version`. You can optionally use a tool like [pyenv](https://github.com/pyenv/pyenv) to manage Python versions on your computer.
+This project targets Python version 3.11.x for best compatibility. Verify the Python3 version that your shell is using with `python3 --version`. You can optionally use a tool like [pyenv](https://github.com/pyenv/pyenv) to manage Python versions on your computer.
```
Comment (patjlm) [inline README.md:222]: asdf was also mentioned elsewhere, do you want to add it here as an alternative to pyenv? And should we introduce in the repo file-based version selection (for both pyenv and asdf?)

Comment (patjlm) [reply README.md:222]:
```
$ git grep '3\.9'
Makefile:       python3.9 -m venv venv
README.md:This project targets Python version 3.9.x for best compatibility. Verify the Python3 version that your shell is using with `python3 --version`. You can optionally use a tool like [pyenv](https://github.com/pyenv/pyenv) to manage Python versions on your computer.
reconcile/utils/terrascript_aws_client.py:            lambda_values["runtime"] = common_values.get("runtime", "python3.9")
setup.py:    python_requires=">=3.9",
setup.py:        "Programming Language :: Python :: 3.9",
```

**Comment by chassing** [comment]
@patjlm, you're way tooooooo fast :) This PR is still WIP. I've always problems to find the "convert to draft" button 🤦

File: README.md (hunk 1 above)
Comment (chassing) [reply README.md]: See [1f761c1](https://github.com/app-sre/qontract-reconcile/pull/3672/commits/1f761c193a23eb855255e3f804df0fbb00907ec3)

**Comment by rporres** [comment]
This is great, thx for tackling this issue ❤️ 

Can you add a quick explanation of the code changes included?

**Comment by chassing** [comment]
@rporres, I updated the PR description. PTAL

**Review by rporres — APPROVED** [review]

**Review by maorfr — APPROVED** [review]

**Pull request merged**